File: app/src/main/java/com/example/waleed/mvi/repositories/github/GitHubRepository.java
package com.example.waleed.mvi.repositories.github;

import com.example.waleed.mvi.pojos.GitHubUser;
import com.example.waleed.mvi.pojos.GitHubUserRepository;

import java.util.List;

import io.reactivex.Observable;

/**
 * Created by waleed on 22/07/2017.
 */

public class GitHubRepository implements GitHubRepositoryDataSource {

    private final GitHubRepositoryDataSource remoteGitHubRepository;

    private static GitHubRepository INSTANCE;


    public static GitHubRepository getInstance(GitHubService gitHubService) {
        if (INSTANCE == null)
            INSTANCE = new GitHubRepository(gitHubService);
        return INSTANCE;
    }

    private GitHubRepository(GitHubService gitHubService) {
        remoteGitHubRepository = new RemoteGitHubRepository(gitHubService);
    }


    @Override
    public Observable<List<GitHubUser>> getUsers() {
        return remoteGitHubRepository.getUsers();
    }

    @Override
    public Observable<List<GitHubUserRepository>> getRepositories(String username) {
        return remoteGitHubRepository.getRepositories(username);
    }
}

File: app/src/main/java/com/example/waleed/mvi/repositories/github/GitHubRepository.kt
package com.example.waleed.mvi.repositories.github

import com.example.waleed.mvi.main.PartialViewState
import com.example.waleed.mvi.pojos.GitHubUser
import com.example.waleed.mvi.pojos.GitHubUserRepository

import io.reactivex.Observable

/**
 * Created by waleed on 22/07/2017.
 */

class GitHubRepository private constructor(val remoteGitHubRepository: GitHubRepositoryDataSource) : GitHubRepositoryDataSource {


    override fun getUsers(): Observable<PartialViewState> {
        return remoteGitHubRepository.getUsers()
    }

    override fun getRepositories(username: String): Observable<List<GitHubUserRepository>> {
        return remoteGitHubRepository.getRepositories(username)
    }

    companion object {

        private var INSTANCE: GitHubRepository? = null


        fun getInstance(remoteGitHubRepository: GitHubRepositoryDataSource): GitHubRepository {
//            if(remoteGitHubRepository !is RemoteGitHubRepository){
//                throw  RuntimeException("Please add RemoteGitHubRepository.class ")
//            }
            if (INSTANCE == null)
                INSTANCE = GitHubRepository(remoteGitHubRepository)
            return INSTANCE as GitHubRepository
        }
    }
}
